### PR TITLE
Init

### PR DESCRIPTION
**Pull request by ajlee21**
This PR initializes an analysis where we are using our simulation approach (ponyo) to identify genes and pathways that are "generic" (i.e. those that have a higher likelihood to be deferentially expressed across a range of different experiments).

This PR is currently pip installing the ponyo module manually until ponyo is uploaded into pypi

The main parts of this PR is the notebook found in ```human_analysis``` which contains a notebook that processes the recount data and trains a VAE.

### CODE REVIEW

File: functions/calc.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    for i in range(num_permutations):
+        
+        sample = gene_rank_df.sample(n=len(gene_rank_df), replace=True)
+        
+        r_perm, p_perm = stats.spearmanr(sample['Rank (simulated)'],
+                            sample['DE_Prior_Rank'])
+        r_perm_values.append(r_perm)
+
+    sort_r_perm_values = sorted(r_perm_values)
+    offset = int(num_permutations*0.025)
```
Comment (ben-heil) [inline functions/calc.py:29]: I would recommend either making .025 a variable to make it more clear what it's doing, or pass the confidence value in as a parameter

File: README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+**Problem**: 
+* This method required a large manual curation effort
+* If you want to perform a new DE analysis in a different biological **context** (i.e. different organism, tissue, media) then you might not have the curated data available. Switch contexts will require a lot of manual effort. 
+* Similarly, using a different statistical method will require re-curation effort
+
+**Question**: Can we use our gene expression simulator to automatically generate null experiments for different contexts in order to overcome manual effort?
+
+## Initial experiment
+**Question**: Can our compendia simulation [link to github] identify the same common enriched pathways as filtered out in GSEA-InContext?
```
Comment (ben-heil) [inline README.md:25]: Should this link to Ponyo?

File: explore_data/nbconverted/0_explore_input_data.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+                      element_blank,
+                      element_text,
+                      element_rect,
+                      element_line,
+                      coords)
+
+from ponyo import utils
+
+from numpy.random import seed
+randomState = 123
```
Comment (ben-heil) [inline explore_data/nbconverted/0_explore_input_data.py:46]: Not a big deal, but `randomState` should be `random_state` to keep everything in snake_case

File: explore_data/nbconverted/0_explore_input_data.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+                 "SRR493967",
+                 "SRR493968",
+                 "SRR493969",
+                 "SRR493970",
+                 "SRR493971",
+                 "SRR493972"]
+#template_data = template_data.loc[smRNA_samples]
+template_data = template_data.drop(smRNA_samples)
+print(template_data.shape)
+template_data.head()"""
```
Comment (ben-heil) [inline explore_data/nbconverted/0_explore_input_data.py:116]: Is this cell supposed to be run, or is it a debug statement? Also do you have a commented out debug statement in your commented out debug statement?

File: explore_data/nbconverted/0_explore_input_data.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    legend_key=element_rect(fill='white', colour='white'), 
+    legend_title=element_text(family='sans-serif', size=15),
+    legend_text=element_text(family='sans-serif', size=12),
+    plot_title=element_text(family='sans-serif', size=15),
+    axis_text=element_text(family='sans-serif', size=12),
+    axis_title=element_text(family='sans-serif', size=15)
+    )
+fig += guides(colour=guide_legend(override_aes={'alpha': 1}))
+#fig += ylim(-750000,10)
+#fig += xlim(-1000000,-750000)
```
Comment (ben-heil) [inline explore_data/nbconverted/0_explore_input_data.py:177]: The debug statements in this cell can be removed

File: explore_data/nbconverted/0_explore_input_data.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+fig += theme(
+    legend_title_align = "center",
+    plot_background=element_rect(fill='white'),
+    legend_key=element_rect(fill='white', colour='white'), 
+    legend_title=element_text(family='sans-serif', size=15),
+    legend_text=element_text(family='sans-serif', size=12),
+    plot_title=element_text(family='sans-serif', size=15),
+    axis_text=element_text(family='sans-serif', size=12),
+    axis_title=element_text(family='sans-serif', size=15)
+    )
```
Comment (ben-heil) [inline explore_data/nbconverted/0_explore_input_data.py:174]: Cool plot theme!

File: functions/download_recount2_data.R [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+## Used R 3.6.3
+#if (!requireNamespace("BiocManager", quietly = TRUE))
+#  install.packages("BiocManager")
+
+#BiocManager::install("recount")
+
+## About
+# recount2 contains over 70,000 uniformly processed human RNA-seq samples spanning
+# TCGA, SRA and GTEx
+# library('recount')
```
Comment (ben-heil) [inline functions/download_recount2_data.R:11]: Don't the functions you're using depend on `recount` being imported?

File: functions/process.py [hunk 8]
```diff
@@ -0,0 +1,70 @@
+import pandas as pd
```
Comment (ben-heil) [inline functions/process.py:1]: Needs a comment about what the file does

File: functions/process.py [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+def replace_ensembl_ids(expression_df,
+gene_id_mapping):
```
Comment (ben-heil) [inline functions/process.py:4]: `gene_id_mapping should either be indented to be level with `expression_df` or the newline between them should be removed

File: functions/process.py [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
```
Comment (ben-heil) [inline functions/process.py:5]: Unnecessary space

File: functions/process.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    # Different cases of many-many mapping between gene ids
+    #count = 0
+    #symbols = []
+    #for symbol, group in gene_id_mapping.groupby("ensembl_gene_id"):
+    #    if group['hgnc_symbol'].nunique() > 1:
+    #        print(group)
+    #        count += 1
+    #        symbols.append(symbol)
+    #count
```
Comment (ben-heil) [inline functions/process.py:28]: Is this commented out code no longer necessary?

File: functions/process.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    # same hgnc symbol. Homologous sequences are paralogous 
+    # if they were separated by a gene duplication event: if a gene in an 
+    # organism is duplicated to occupy two different positions in the same 
+    # genome, then the two copies are paralogous
+    
+    # Remove paralogs
+
+    gene_expression = gene_expression.iloc[:,~gene_expression.columns.str.contains("PAR_Y")]
+
+
```
Comment (ben-heil) [inline functions/process.py:40]: Should be a single space not a double (sorry for all the style comments)

File: functions/process.py [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+    # Replace ensembl ids with gene symbol
+    gene_expression.columns = gene_expression.columns.map(gene_id_mapping['hgnc_symbol'])
+
+    # Remove rows where we couldn't map ensembl id to gene symbol
+    gene_expression = gene_expression.iloc[:,gene_expression.columns != ""]
+    gene_expression = gene_expression.iloc[:,gene_expression.columns.notnull()]
+
+
+    # Save
+    #gene_expression.t.to_csv(expression_file, float_format='%.5f', sep='\t')
```
Comment (ben-heil) [inline functions/process.py:69]: Should this commented out code be removed?

File: functions/process_names.R [hunk 14]
```diff
@@ -0,0 +1,28 @@
+
```
Comment (ben-heil) [inline functions/process_names.R:1]: Can you add a comment about what this file does?

File: functions/process_names.R [hunk 15]
```diff
@@ -0,0 +1,28 @@
+  
+## Run this once to setup environment
+## Used R 3.6.3
+#if (!requireNamespace("BiocManager", quietly = TRUE))
+#  install.packages("BiocManager")
+
+convertEnsembl2Symbol <- function(ensembl.genes) {
```
Comment (ben-heil) [inline functions/process_names.R:7]: Should be `convert_ensembl_to_symbol` to be consistent with your other R code

File: human_analysis/nbconverted/1_process_recount2_data.py [hunk 16]
```diff
@@ -0,0 +1,389 @@
...
+import sys
+import pandas as pd
+import numpy as np
+import random
+import rpy2
+import seaborn as sns
+from sklearn import preprocessing
+import pickle
+
+sys.path.append("../")
```
Comment (ben-heil) [inline human_analysis/nbconverted/1_process_recount2_data.py:28]: If you add a setup.py file to your repo, you can install the repo as a module and not have to manually set the path to your python functions. Probably would be the subject of a different PR if you choose to do so though

File: human_analysis/nbconverted/1_process_recount2_data.py [hunk 17]
```diff
@@ -0,0 +1,389 @@
...
+
+
+# ### Download subset of recount2 to use as a compendium
+# The compendium will be composed of random experiments + the selected template experiment
+
+# In[4]:
+
+
+get_ipython().run_cell_magic('R', '', '# Select 59\n# Select a\n# Run one time\n#if (!requireNamespace("BiocManager", quietly = TRUE))\n#    install.packages("BiocManager")\n#BiocManager::install("derfinder")\n#BiocManager::install("recount")')
+
```
Comment (ben-heil) [inline human_analysis/nbconverted/1_process_recount2_data.py:69]: This cell should probably be uncommented for the sake of future users of the notebook

File: human_analysis/nbconverted/1_process_recount2_data.py [hunk 18]
```diff
@@ -0,0 +1,389 @@
...
+
+# In[6]:
+
+
+#%%R -i project_id -i local_dir
+
+#source('../functions/download_recount2_data.R')
+
+#get_recount2_compendium(project_id, local_dir)
```
Comment (ben-heil) [inline human_analysis/nbconverted/1_process_recount2_data.py:84]: Isn't this cell a duplicate of the following cell?

File: human_analysis/nbconverted/1_process_recount2_data.py [hunk 19]
```diff
@@ -0,0 +1,389 @@
...
+print(original_compendium.shape)
+original_compendium.head()
+
+
+# In[28]:
+
+
+# Replace ensembl ids with gene symbols
+#original_compendium = process.replace_ensembl_ids(original_compendium,
+#                                                gene_id_mapping)
```
Comment (ben-heil) [inline human_analysis/nbconverted/1_process_recount2_data.py:313]: Can be removed

File: functions/download_recount2_data.R [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+  selected_project_ids <- append(selected_project_ids, template_project_id)
+
+  ## Data
+
+  # Get data associated with project ids
+  # Download the RangedSummarizedExperiment object at the gene level for 
+  for(i in 1:length(selected_project_ids)){
+    if(!file.exists(file.path(selected_project_ids[i], 'rse_gene.Rdata'))) {
+      download_study(selected_project_ids[i])
+      load(file.path(selected_project_ids[i], 'rse_gene.Rdata'), verbose = TRUE)
```
Comment (jjc2718) [inline functions/download_recount2_data.R:91]: I think you can just put the `load` outside the file check, rather than repeating this line in the else block. Something like this:
```R
 if (!file.exists(file.path(selected_project_ids[i], 'rse_gene.Rdata'))) {
      download_study(selected_project_ids[i])
}
load(file.path(selected_project_ids[i], 'rse_gene.Rdata'), verbose = TRUE)
```

File: functions/calc.py (hunk 1 above)
Comment (ajlee21) [reply functions/calc.py:29]: good point I will add this

File: README.md (hunk 2 above)
Comment (ajlee21) [reply README.md:25]: Yes! That was a placeholder I forgot to replace

File: functions/process.py [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+    # Remove paralogs
+
+    gene_expression = gene_expression.iloc[:,~gene_expression.columns.str.contains("PAR_Y")]
+
+
+    # Case 2: Same ensembl ids are mapped to different gene symbol twice (CCL3L1, CCL3L3)
+    # ENSG00000187510.7  ENSG00000187510    C12orf74  
+    # ENSG00000187510.7  ENSG00000187510     PLEKHG7  
+    # Manually map based on what is found on ensembl site
+    manual_mapping = dict({'ENSG00000187510.7':'PLEKHG7',
```
Comment (jjc2718) [inline functions/process.py:45]: I think the call to `dict` is redundant here - Python should parse what's in the curly brackets as a dict by default:
```python
my_dict = {'a': 1}
type(my_dict) # dict
type(dict(my_dict)) # dict
```

File: explore_data/nbconverted/0_explore_input_data.py (hunk 4 above)
Comment (ajlee21) [reply explore_data/nbconverted/0_explore_input_data.py:116]: This cell is supposed to run and was commented out while debugging.

File: functions/download_recount2_data.R (hunk 7 above)
Comment (ajlee21) [reply functions/download_recount2_data.R:11]: Recount is actually imported in the notebook. I assume either would work...

File: README.md [hunk 22]
```diff
@@ -0,0 +1,45 @@
...
+
+Processing and analysis scripts were performed using the conda environment specified in `environment.yml`.
+To build and activate this environment run:
+
+```bash
+# conda version 4.6.12
+conda env create -f environment.yml
+
+conda activate generic_expression
+```
```
Comment (jjc2718) [inline README.md:45]: You might also consider adding some documentation on how to set up the config for a new system (i.e. just briefly mention that the home directory is set in the `config_human.tsv` file, and to run it you'll need to change those filepaths to match the directory the repo is installed in).

I'm not sure how much you're expecting other people to use this repo (as opposed to just looking at the results), but it's still probably good to document it a bit.

File: functions/download_recount2_data.R [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+    # Counts are the number of reads that map that each gene
+    
+    # Scale counts by total coverage (total reads across the genome) per sample
+    # Mixed paired-end and single-end reads per run
+    # Most runs are single-end
+    # RPKM: normalizes for sequencing depth (coverge per gene) and gene length. Usually used for single-end
+    rse_rpkm <- getRPKM(rse_gene, length_var="bp_length")
+    
+    # Concatenate scaled counts into one matrix
+    if (i==1){
```
Comment (jjc2718) [inline functions/download_recount2_data.R:106]: Make sure to be consistent with the spacing of your `if`/`else` statements in these R scripts (I think the standard is a space after the `if` and before the `(`, and a space after the `)` and before the `{`).

File: functions/process.py (hunk 11 above)
Comment (ajlee21) [reply functions/process.py:28]: This code isn't being used no, it was more of a note to myself of how I discovered the different cases that needed to be addressed when mapping gene ids

File: human_analysis/nbconverted/1_process_recount2_data.py (hunk 17 above)
Comment (ajlee21) [reply human_analysis/nbconverted/1_process_recount2_data.py:69]: Eventually I will need to uncomment this out. For now its commented so that I don't have to re-download a new batch of recount2 data each time i re-run this notebook for testing

File: human_analysis/nbconverted/1_process_recount2_data.py (hunk 18 above)
Comment (ajlee21) [reply human_analysis/nbconverted/1_process_recount2_data.py:84]: One cell is getting a compendia of recount2 data, the one below is getting a specific experiment.

File: human_analysis/nbconverted/1_process_recount2_data.py (hunk 19 above)
Comment (ajlee21) [reply human_analysis/nbconverted/1_process_recount2_data.py:313]: This is another comment that is temporarily commented for the purposes of debugging but will need to be uncommented later

File: config_human.tsv [hunk 24]
```diff
@@ -0,0 +1,17 @@
+local_dir	"/home/alexandra/Documents/Data/Generic_expression_patterns/"
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply config_human.tsv:1]: I am using this in both R and python functions.

Interesting idea...perhaps we can chat offline about this. I'm not sure if I'm exactly understanding the benefit compared to using it in this config file...so i'm sure i'm missing something here

File: functions/download_recount2_data.R (hunk 20 above)
Comment (ajlee21) [reply functions/download_recount2_data.R:91]: Good catch

File: config_human.tsv (hunk 24 above)
[the remark this answers is not in the document]
Comment (jjc2718) [reply config_human.tsv:1]: Ok, I think if you're planning to use this in your R scripts too, what you have is fine.

In your config file you repeat `/home/alexandra/Documents/Data/Generic_expression_patterns/` in lots of places - this kind of repetition just makes me nervous in general, since you have to remember to change it in all of those places if you move the repo to a new directory. Just trying to think of ways to avoid that and have a single point of truth for where the repo is located, but I don't know of a good way to do that in a text file.

Happy to chat offline if this is still unclear.

[the remark this answers is not in the document]
Comment (ajlee21) [reply config_human.tsv:1]: Ah..I see now. I think this config file was a nice step up from previously me defining this dir in each separate notebook. So at least the change is happening within this one config file. But I agree that it would be even better if there is a way to have no repetition within that file too. For now I think I'll leave it, but keep it in mind moving forward for sure
